libcloud/compute/drivers/vsphere.py

```diff
@@ -12,543 +12,216 @@
-
-    @wrap_non_libcloud_exceptions
-    def ex_get_hosts(self):
-        """
-        Return all the available hosts.
-
-        :rtype: ``dict``
-        """
-        result = self.connection.client.get_hosts()
-        return result
-
-    @wrap_non_libcloud_exceptions
-    def ex_get_datastores(self):
-        """
-        Return all the available datastores.
-
-        :rtype: ``dict``
-        """
-        result = self.connection.client.get_datastores()
-        return result
-
-    @wrap_non_libcloud_exceptions
-    def ex_get_node_by_path(self, path):
-        """
-        Retrieve Node object for a VM with a provided path.
-
-        :type path: ``str``
-        :rtype: :class:`libcloud.compute.base.Node`
-        """
-        vm = self.connection.client.get_vm_by_path(path)
-        node = self._to_node(vm=vm)
-        return node
+        extra = {
+            "path": config.vmPathName,
+            "operating_system": config.guestFullName,
+            "os_type": "windows" if "Microsoft" in config.guestFullName \
+                       else "unix",
+            "memory_MB": config.memorySizeMB,
+            "cpus": config.numCpu,
+            "overallStatus": str(vm.summary.overallStatus),
+            "datacenter": datacenter,
+        }
 
-    def ex_get_node_by_uuid(self, uuid):
-        """
-        Retrieve Node object for a VM with a provided uuid.
+        kwargs = {
+            "id": config.instanceUuid,
+            "name": config.name,
+            "state": self.NODE_STATE_MAP.get(vm.summary.runtime.powerState,
+                                             NodeState.UNKNOWN),
+            "public_ips": [],
+            "private_ips": [],
+            "driver": self,
+            "extra": extra,
+        }
 
-        :type uuid: ``str``
-        """
-        vm = self._get_vm_for_uuid(uuid=uuid)
-        node = self._to_node(vm=vm)
+        for nic in vm.guest.net:
+            for addr in nic.ipAddress:
+                try:
+                    if is_public_subnet(addr):
+                        kwargs['public_ips'].append(addr)
+                    else:
+                        kwargs['private_ips'].append(addr)
+                except OSError:  # inet_aton error for ipv6 addrs
+                    pass
+
+        node = Node(**kwargs)
+        node._uuid = kwargs['id']
         return node
 
-    @wrap_non_libcloud_exceptions
-    def ex_get_server_type(self):
-        """
-        Return VMware installation type.
-
-        :rtype: ``str``
-        """
-        return self.connection.client.get_server_type()
-
-    @wrap_non_libcloud_exceptions
-    def ex_get_api_version(self):
-        """
-        Return API version of the vmware provider.
-
-        :rtype: ``str``
-        """
-        return self.connection.client.get_api_version()
+    def list_locations(self):
+        """ Return the clusters available on this vSphere instance """
+        return [NodeLocation(id="{}/{}".format(dc.name, cl.name),
```

Comment:
Upstream still supports python 2.6, so you can't reply in implicit ordering of format strings. Either:

- `"{0}/{1}".format(dc.name, cl.name)`
- `'/'.join((dc.name, cl.name))`